src/index.ts

```diff
@@ -0,0 +1,271 @@
+  }
+}
+
+export async function autoReplication(
+  infoServerAddress: string,
+  serverName: string,
+  apiKey: string
+): Promise<void> {
+  const uri = `https://${infoServerAddress}/v1/edgeServersInternal/${serverName}s?apiKey=${apiKey}`
+  try {
+    const result = await fetch(uri, {
+      method: 'GET'
+    })
+    const resultObj = await result.json()
+    for (const cluster of resultObj.clusters) {
+      for (const serverUri in cluster.servers) {
+        if (
+          typeof cluster.servers[serverUri].username === 'string' &&
+          typeof cluster.servers[serverUri].password === 'string'
+        ) {
+          console.log(
+            `Username and Password for server ${serverUri} exists, attempting replication between servers.`
+          )
+          const sourceUri = `https://${cluster.servers[serverUri].username}:${cluster.servers[serverUri].password}@${serverUri}:6984`
+          const destinationUri = config.couchDbFullpath
+          const existingDbsUri = `${sourceUri}/_all_dbs`
+          let finalDbList
+          try {
+            const existingDbsQuery = await fetch(existingDbsUri, {
+              method: 'GET'
+            })
+            const existingDbs = await existingDbsQuery.json()
+            finalDbList = existingDbs.filter(
+              dbName => dbName !== '_users' || dbName !== '_replicator'
+            )
+          } catch {
+            console.log(
+              `Could not get list of existing dbs for ${serverUri}, continuing other replications`
+            )
+            continue
+          }
+          for (const db of finalDbList) {
+            await dbReplication(sourceUri, db, destinationUri, db, true)
+          }
+        } else {
+          console.log(
+            `Username and Password for server ${serverUri} does not exist, cannot attempt replication.`
+          )
+        }
+      }
+    }
+  } catch {
+    console.log('Replication between servers failed.')
+  }
+}
+export async function nodeJsClustering(
+  mainFunction: Function,
+  couchDbFullPath: string,
+  couchSchema: any,
+  serverName: string,
+  infoServerAddress: string,
+  infoServerApiKey: string,
+  instanceCount?: number
+) {
+  const numCPUs = cpus().length
+  if (cluster.isMaster) {
+    prepareCouch(couchDbFullPath, couchSchema)
+      .then(async () =>
+        autoReplication(infoServerAddress, serverName, infoServerApiKey)
+      )
+      .then(() => {
+        const instances = instanceCount ?? numCPUs
+
+        // Fork workers.
+        for (let i = 0; i < instances; i++) {
+          cluster.fork()
+        }
+
+        // Restart workers when they exit
+        cluster.on('exit', (worker, code, signal) => {
+          console.log(
+            `Worker ${worker.process.pid} died with code ${code} and signal ${signal}`
+          )
+          console.log(`Forking new worker process...`)
+          cluster.fork()
+        })
+      })
+      .catch(failStartup)
+  } else {
+    mainFunction()
+  }
+}
```

Comment:
If I remember correctly, the proposed clustering function's interface was something like `forkProcess(instanceCount?: number): void`. It was only meant to call `cluster.fork()` and attach an exit event listener to restart workers. All the other things such as preparing couch and setting up replication would be left up to the application to decide to do or not.

Example of the usage:

```ts
if (cluster.isMaster) {
  await prepareCouch(couchDbPath, couchSchema)
  await autoReplication(...)
  forkProcess()
}
else {
  await main()
}
```
```